loki/transformations/transform_loop.py

```diff
@@ -608,3 +610,133 @@ def loop_fission(routine, promote=True, warn_loop_carries=True):
+
+            acc = functools.reduce(op.add,
+                                   [
+                                       # Create a copy of the loop body for every value of the iterator
+                                       SubstituteExpressions({o.variable: sym.IntLiteral(i)}).visit(o.body)
+                                       for i in unroll_range
+                                   ],
+                                   ())
+
+            if depth is None or depth >= 1:
+                acc = [self.visit(a, depth=depth, warn_iterations_length=warn_iterations_length) for a in acc]
+
+            return as_tuple(flatten(acc))
+        else:
+            return Loop(
+                variable=o.variable,
+                body=self.visit(o.body, depth=depth, warn_iterations_length=warn_iterations_length),
+                bounds=o.bounds
+                        )
+
+
+def loop_unroll(routine):
+    """
+    Search for ``!$loki loop-unroll`` pragmas in loops and unroll them.
+
+    The expected pragma syntax is
+    ``!$loki loop-fission [depth(n)]``
+    where ``depth(n)`` controls the unrolling of nested loops. For instance,
+    ``depth(1)`` will only unroll the top most loop of a set of nested loops.
+    However, a child nested loop with a more restrictive depth will not be
+    able to override its parent's depth. If ``depth(n)`` is not specified,
+    then all loops nested under a parent with this pragma will be unrolled.
+    E.g. The code sample below will only unroll A and B, but not C:
+
+    ! Loop A
+    !$loki loop-unroll depth(1)
+    DO a = 1, 10
+        ! Loop B
+        !$loki loop-unroll
+        DO b = 1, 10
+            ...
+        END DO
+        ! Loop C - will not be unrolled
+        DO c = 1, 10
+            ...
+        END DO
+    END DO
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine in which loop unrolling is to be applied.
+    """
+
+    class PragmaLoopUnrollTransformer(Transformer):
+        def visit_Loop(self, o, *args, **kwargs):
+            # Check for pragmas
+            if is_loki_pragma(o.pragma, starts_with='loop-unroll'):
+                parameters = get_pragma_parameters(o.pragma, starts_with='loop-unroll')
+
+                # Get the depth
+                param = parameters.get('depth', None)
+                depth = int(param) if param is not None else None
+
+                # Unroll and recurse
+                unrolled_loop = LoopUnrollTransformer().visit(o, depth=depth)
+
+                # unrolled_loop could be either an unrollable Loop() or a Tuple() of Nodes
+                try:
+                    return as_tuple(flatten([self.visit(a) for a in as_tuple(flatten(unrolled_loop))]))
+                # Loop() is not iterable
+                except TypeError:
+                    return self.visit(unrolled_loop)
+            else:
+                return Loop(variable=o.variable, body=self.visit(o.body), bounds=o.bounds)
```

Comment:
I'll implement the style changes (clearly I need to enable pylint on my IDE!). However, `super()` doesn't implement `visit_Loop()` and using `visit()` instead could cause infinite recursion as there'd be no base case for a `Loop` node without an associated pragma.

The intent of this `else` branch was to recurse into the body of a loop which wasn't going to be unrolled, in case it contained loops that did need to be unrolled. Initially I went with something like the below, but `Loop` nodes seem to be immutable (or at least that field is).

```
unrolled_body = self.visit(o.body)
o.body = unrolled_body
return o
```

What's the best approach for this?

Reply:
Right, it should probably be `super().visit_Node(o, *args, **kwargs)`. That will make sure the other options a Transformer may have remain active.

Otherwise, yes, IR nodes are immutable by default, but there is a hidden "update" mechanism for in-place rebuild, you could unlock this by doing
```py
o._update(body=self.visit(o.body, **kwargs))
return o
```